src/main/java/io/javalin/core/PathParser.kt

```diff
@@ -8,30 +8,62 @@ package io.javalin.core
 
 import io.javalin.http.util.ContextUtil
 
-class PathParser(
-        path: String,
-        private val pathParamNames: List<String> = path.split("/")
-                .filter { it.startsWith(":") }
-                .map { it.replace(":", "") },
-        private val matchRegex: Regex = pathParamNames
-                .fold(path) { p, name -> p.replace(":$name", "[^/]+?") } // Replace path param names with wildcards (accepting everything except slash)
-                .replace("//", "/") // Replace double slash occurrences
-                .replace("/*/", "/.*?/") // Replace star between slashes with wildcard
-                .replace("^\\*".toRegex(), ".*?") // Replace star in the beginning of path with wildcard (allow paths like (*/path/)
-                .replace("/*", "/.*?") // Replace star in the end of string with wildcard
-                .replace("/$".toRegex(), "/?") // Replace trailing slash with optional one
-                .run { if (!endsWith("/?")) this + "/?" else this } // Add slash if doesn't have one
-                .run { "^" + this + "$" } // Let the matcher know that it is the whole path
-                .toRegex(),
-        private val pathParamRegex: Regex = matchRegex.pattern.replace("[^/]+?", "([^/]+?)").toRegex(RegexOption.IGNORE_CASE)) {
-
-    fun matches(url: String) = url matches matchRegex
-
-    fun extractPathParams(url: String) = pathParamNames.zip(values(pathParamRegex, url)) { name, value ->
-        name to ContextUtil.urlDecode(value)
-    }.toMap()
+class PathParser(path: String) {
+    val parsedPath = ParsedPath.fromPath(path)
+
+    fun matches(url: String): Boolean = url matches parsedPath.matchRegex
+
+    fun extractPathParams(url: String) = parsedPath.pathParamNames
+            .zip(values(parsedPath.pathParamRegex, url)) { name, value ->
+                name to ContextUtil.urlDecode(value)
+            }.toMap()
 
     // Match and group values, then drop first element (the input string)
     private fun values(regex: Regex, url: String) = regex.matchEntire(url)?.groupValues?.drop(1) ?: emptyList()
+}
+
+data class ParsedPath(val segments: List<PathSegment>) {
+    companion object {
+        fun fromPath(path: String): ParsedPath {
+            val segments = path.split("/")
+                    .filter { it.isNotEmpty() }
+                    .map {
+                        when {
+                            it.startsWith(":") -> PathSegment.Parameter(it.replace(":", ""))
```

Comment:
```suggestion
                            it.startsWith(":") -> PathSegment.Parameter(it.removePrefix(":"))
```

Reply:
> Is this equivalent in performance to the current solution?

I'm going to make some performance tests. But I think it will be probably a little bit worse as there is more logic. But as the parsing happens on the initialisation of the app, the runtime performance should be not affected. If it is slower, just the startup time is affected.

Reply:
I don't think the startup performance matters really, just wondering if the resulting regex is equivalent.

Reply:
Here is my performance test. Please note that I have no experience in performance testing, so this is probably far from perfect.

First the code
```kotlin
package io.javalin.core

import io.javalin.http.util.ContextUtil
import org.junit.Test

class PathParserTest {
    val NUMBER_OF_RUNS = 100
    val NUMBER_OF_URLS = 100_000
    val testUrl = "/this/is/a/long/url/:param/*"

    fun stopTime(callback: () -> Unit): Long {
        val start = System.currentTimeMillis()
        callback()
        return System.currentTimeMillis() - start
    }

    fun runTests(compute: () -> Unit) {
        val allTimes = mutableListOf<Long>()
        repeat(NUMBER_OF_RUNS) {
            val time = stopTime {
                repeat(NUMBER_OF_URLS) {
                    compute()
                }
            }
            allTimes.add(time)
        }
        println("Total Time ${allTimes.sum()}")
        println("Average Time per run ${allTimes.sum() / NUMBER_OF_RUNS}")
    }

    fun runStartupTests() {
        println("-- Startup")
        println("- OLD")
        runTests { val parser = OldPathParser(testUrl) }
        println("- NEW")
        runTests { val parser = PathParser(testUrl) }
    }

    fun runRuntimeTests() {
        println("-- Runtime")
        println("- OLD")
        val old = OldPathParser(testUrl)
        runTests {
            old.matches("/some/url")
            old.matches(testUrl)
        }
        println("- NEW")
        val new = PathParser(testUrl)
        runTests {
            new.matches("/some/url")
            new.matches(testUrl)
        }
    }

    fun runTests() {
        runStartupTests()
        runRuntimeTests()
    }

    @Test
    fun performPerformanceTests() {
        println("--- First time --")
        runTests()
        println("--- A second time for reference --")
        runTests()
        println("--- A third time for reference --")
        runTests()
    }
}
```

and the results:

```
--- First time --
-- Startup
- OLD
Total Time 27332
Average Time per run 273
- NEW
Total Time 17038
Average Time per run 170
-- Runtime
- OLD
Total Time 2858
Average Time per run 28
- NEW
Total Time 2440
Average Time per run 24

--- A second time for reference --
-- Startup
- OLD
Total Time 26564
Average Time per run 265
- NEW
Total Time 16951
Average Time per run 169
-- Runtime
- OLD
Total Time 2460
Average Time per run 24
- NEW
Total Time 2456
Average Time per run 24

--- A third time for reference --
-- Startup
- OLD
Total Time 26807
Average Time per run 268
- NEW
Total Time 17045
Average Time per run 170
-- Runtime
- OLD
Total Time 2456
Average Time per run 24
- NEW
Total Time 2479
Average Time per run 24
```

I did three runs to counter possible optimization by the jvm.

It seems that the startup time has even improved. Probably all these replacements not as efficient as thought. The runtime performance seems to be same.

Reply:
Great, thanks for being thorough! I'm no expert on performance either, but there are a couple of performance tests in `src/test/java/io/javalin/performance` you could look at if you need to measure something in the future. I think the correct way to do performance tests is to use JMH, but this approach still gives an indication.